### PR TITLE
(WIP) feat/core: implement network event observer

### PR DESCRIPTION
**Pull request by madadam**
Note: **DO NOT MERGE** yet. Missing tests.

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @krishnaIndia

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/core/client/mod.rs [hunk 1]
```diff
@@ -39,14 +40,18 @@ use self::mock_routing::MockRouting as Routing;
...
 use std::rc::Rc;
-use std::sync::mpsc::{self, Receiver};
+use std::sync::mpsc;
 use std::time::Duration;
 
 const CONNECTION_TIMEOUT_SECS: u64 = 60;
 const ACC_PKT_TIMEOUT_SECS: u64 = 60;
 const IMMUT_DATA_CACHE_SIZE: usize = 300;
 
+pub type NetworkEventSender = stream::Sender<NetworkEvent, ()>;
```
Comment (ustulation) [inline src/core/client/mod.rs]: We have the `Tx/Rx` convention, why are we using the longer alternative everywhere - `sender/receiver` ?

File: src/core/client/mod.rs [hunk 2]
```diff
@@ -59,7 +64,8 @@ pub struct Client {
 
 struct Inner {
     routing: Routing,
-    heads: HashMap<MessageId, Complete<CoreEvent>>,
+    core_event_completes: HashMap<MessageId, Complete<CoreEvent>>,
+    network_event_senders: Vec<NetworkEventSender>,
```
Comment (ustulation) [inline src/core/client/mod.rs]: why is this a vector ?

File: src/core/client/routing_el.rs [hunk 3]
```diff
@@ -104,15 +111,35 @@ pub fn parse_mutation_err(reason_raw: &[u8]) -> MutationError {
...
+        if let Some(complete) = complete {
+            complete.complete(event);
         }
         None
     });
 
     core_tx.send(msg).is_ok()
 }
+
+fn fire_network_event(core_tx: &CoreMsgTx, event: NetworkEvent) {
```
Comment (ustulation) [inline src/core/client/routing_el.rs]: Seems complicated for a simple task

File: src/core/event.rs [hunk 4]
```diff
@@ -45,12 +45,12 @@ pub enum _NetworkEvent {
     Terminated,
```
Comment (ustulation) [inline src/core/event.rs]: This variant may no longer be required as there is no `rx` as such.

**Comment by ustulation** [comment]
Discarding in favour of https://github.com/maidsafe/safe_core/pull/365

**Pull request closed without merging**